In [1]:
%run ./py_templates/1_Boilerplate_Numeric.py

Author        : sm626967
Python        : 3.7.4 
NumPy         : 1.18.1
Encoding      : utf-8
Timestamp     : 1595687955.346698


Use the following schematic for reference:

![Differentiator Circuit](diff_ref_ckt.JPG).

There is an inverting opamp, a feedback resistor, and two series elements, $C_1$ and $R_2$ which are in series with the input signal. The circuit is based on the capacitor's current to voltage relationship, $$i_C=C_1 \cdot \frac{dv}{dt}$$

where $i_C$ is the current through the capacitor, C is the capacitance of the capacitor, and v is the voltage across the capacitor. The current flowing through the capacitor is then proportional to the derivative of the voltage across the capacitor. This current can then be connected to a resistor, which has the current to voltage relationship $$I=\frac{v}{R_1}$$

The op-amp input has a very high input impedance (it also forms a virtual ground due to the presence of negative feedback), so the entire input current has to flow through $R_1$

If we let $V_{out}$ represent the voltage across the resistor and $V_{in}$ is the voltage across the capacitor, we can rearrange these two equations to obtain the following equation: 

$$v_{out}=-R_1C_1 \cdot \frac{dv_{in}}{dt}$$

* Output is proportional to the rate of change of the input. So, the op amp acts as a differentiator.
* Above equation is true for any frequency signal.
* The negative sign indicates that there is 180° phase shift in the output with respect to the input,

It can be shown that in an ideal situation the voltage across the resistor will be proportional to the derivative (rate of change) of the voltage across the capacitor with a gain of RC. 

There are some limitations to the performance of this circuit as a differentiator. The most notable limitation is that at high frequencies (defined soon):

* this simple differentiator circuit becomes unstable and oscillates
* the circuit becomes sensitive to noise: when amplified, noise dominates the input signal.

To mitigate against oscillation, another capacitor of a small value may be placed in parallel with the feedback resistor $R_1$

This provides us with the following frequency relationship(s) based upon $RC$ time constants:

Let's define a few terms first: <br><br>
$$f_1=\frac{1}{2 \pi R_1 C}$$<br><br>
$f_1$ is the frequency at and below this value causing the circuit to behave like a differentiator.

$$f_2=\frac{1}{2 \pi R C_2}$$<br><br>

Note that $C_2$ is not illustrated in the figure at the top of this notebook. It is the small value capcitor placed in parallel with the feedback resistor $R_1$ to mitigate against oscillation.

In any case, frequencies above $f_2$ will cause the circuit to behave as an integrator.

Frequencies between $f_1$ and $f_2$ will cause the circuit to behave like a very low gain amplifier or a buffer. 

What happens if you don't have $C_2$ in parallel with $R_1$, <code>OR</code> if $C_1$=$C_2$ and $R_1$=$R_2$?

In this case, any frequencies above $f_1$ will cause the circuit to act as an integrator, and $$f=\frac{1}{2 \pi RC}$$

Let's calculate the $f_1$ of the circuit shown at the top of this notebook:

In [2]:
C=0.0047E-6
R1=15.0E3
f1=np.reciprocal(2*pi*R1*C)
print("f_1: "'{0:.5f}'.format(f1),"Hz")

f_1: 2257.51692 Hz


Meaning, that in this case and using the values shown in the previous cell, any input frequencies at or below <code>2257.51692 Hz</code> will cause this circuit to behave like a differentiator, but frequencies above this frequency will cause the circuit to act as an integrator. A small capacitor in parallel with $R_1$ would mitigate against oscillations and move the boundary condition for integrator behaviour up in frequency space.